batch.js

```diff
@@ -0,0 +1,317 @@
+const debugFactory = require('debug')
+
+const debug = debugFactory('graphile-upsert-batch')
+
+function PgMutationUpsertBatchPlugin(builder) {
+  builder.hook('GraphQLObjectType:fields', (fields, build, context) => {
+    const {
+      extend,
+      newWithHooks,
+      parseResolveInfo,
+      pgIntrospectionResultsByKind,
+      pgGetGqlTypeByTypeIdAndModifier,
+      pgGetGqlInputTypeByTypeIdAndModifier,
+      pgSql: sql,
+      gql2pg,
+      graphql: {
+        GraphQLObjectType,
+        GraphQLList,
+        GraphQLInputObjectType,
+        GraphQLNonNull,
+        GraphQLString,
+      },
+      pgColumnFilter,
+      inflection,
+      pgQueryFromResolveData: queryFromResolveData,
+      pgOmit: omit,
+      pgViaTemporaryTable: viaTemporaryTable,
+      describePgEntity,
+      sqlCommentByAddingTags,
+      pgField,
+    } = build
+    const {
+      scope: { isRootMutation },
+      fieldWithHooks,
+    } = context
+    if (!isRootMutation) {
+      return fields
+    }
+
+    return extend(
+      fields,
+      pgIntrospectionResultsByKind.class
+        .filter(table => !!table.namespace)
+        .filter(table => table.isSelectable)
+        .filter(table => table.isInsertable && !omit(table, 'create')) // we ignore update omits for simplicity
+        .reduce((memo, table) => {
+          const Table = pgGetGqlTypeByTypeIdAndModifier(table.type.id, null)
+          if (!Table) {
+            debug(
+              `There was no table type for table '${table.namespace.name}.${
+                table.name
+              }', so we're not generating an upsert batch mutation for it.`,
+            )
+            return memo
+          }
+          const TableInput = pgGetGqlInputTypeByTypeIdAndModifier(
+            table.type.id,
+            null,
+          )
+          if (!TableInput) {
+            debug(
+              `There was no input type for table '${table.namespace.name}.${
+                table.name
+              }', so we're going to omit it from the upsert batch mutation.`,
+            )
+          }
+          const tableTypeName = inflection.tableType(table)
+          const inputFieldName = inflection.pluralize(
+            inflection.tableFieldName(table),
+          )
+          const InputType = newWithHooks(
+            GraphQLInputObjectType,
+            {
+              name: `Upsert${tableTypeName}BatchInput`,
```

Comment:
Moving this to an inflector would enable others to customise it without forking. You can use makeAddInflectorsPlugin and makePluginByCombiningPlugins to achieve this should you so wish. (You can do this in a later version.)

Reply:
[bookmarked](https://github.com/micimize/graphile-upsert-plugin/issues/1) this to come back to in a future version